lib/bolt/pal/yaml_plan.rb

```diff
@@ -3,6 +3,35 @@
 module Bolt
   class PAL
     class YamlPlan
+      PARAMETER_KEYS = Set['type', 'default']
+      COMMON_STEP_KEYS = %w[name description target].freeze
+      STEP_KEYS = {
+        'command' => {
+          'allowed_keys' => Set['command'].merge(COMMON_STEP_KEYS),
+          'required_keys' => Set['target']
+        },
+        'script' => {
+          'allowed_keys' => Set['script', 'parameters'].merge(COMMON_STEP_KEYS),
+          'required_keys' => Set['target']
+        },
+        'task' => {
+          'allowed_keys' => Set['task', 'parameters'].merge(COMMON_STEP_KEYS),
+          'required_keys' => Set['target']
+        },
+        'plan' => {
+          'allowed_keys' => Set['plan', 'parameters'].merge(COMMON_STEP_KEYS),
+          'required_keys' => Set.new
+        },
+        'source' => {
```

Comment:
The upload step is the only step that does not have a single key that can identify it (must have both `source` and `destination`). As WIP i ignore this and identify it on `source`. Will need to update the logic.

Reply:
we could keep this inner hash separate and have both source and destination point to it? That may be more complex in other ways but it seems like we should give the same error regardless of which special key is missing.

Reply:
Went with having source and destination map to the a copy of a hash with the same info. The only added complexity I can tell it adds validating only a single action is specified per step.